pkg/kubelet/kubelet_stats.go

```diff
@@ -0,0 +1,644 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubelet
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	cadvisorapiv1 "github.com/google/cadvisor/info/v1"
+	cadvisorapiv2 "github.com/google/cadvisor/info/v2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	internalapi "k8s.io/kubernetes/pkg/kubelet/apis/cri"
+	statsapi "k8s.io/kubernetes/pkg/kubelet/apis/stats/v1alpha1"
+	"k8s.io/kubernetes/pkg/kubelet/cadvisor"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	"k8s.io/kubernetes/pkg/kubelet/leaky"
+	"k8s.io/kubernetes/pkg/kubelet/network"
+	"k8s.io/kubernetes/pkg/kubelet/server/stats"
+	kubetypes "k8s.io/kubernetes/pkg/kubelet/types"
+)
+
+// containerStatsProvider is an interface for extracting container stats.
+type containerStatsProvider interface {
+	ListPodStats() ([]statsapi.PodStats, error)
+	ImageFsStats() (*statsapi.FsStats, error)
+}
+
+var _ containerStatsProvider = &criStatsProvider{}
+var _ containerStatsProvider = &cadvisorStatsProvider{}
+
+// criStatsProvider implements the containerStatsProvider interface by getting
+// the container stats from CRI.
+type criStatsProvider struct {
+	runtimeService   internalapi.RuntimeService
+	imageService     internalapi.ImageManagerService
+	cadvisor         cadvisor.Interface
```

Comment:
Each container filesystem stats contain the capacity/available bytes/inodes. We need to get those metrics from cadvisor. This is duplicated information for each container - it's available as the node level stats as well. Can we just not provide these stats in ContainerStats?